### PR TITLE
Optimization for train_test_split indices shuffling

### PR DESCRIPTION
**Pull request by Alexsandruss**
# Description
Optimization of indices shuffling for scikit-learn train_test_split function.

### CODE REVIEW

**Comment by Alexsandruss** [comment]
@PetrovKP, review, please.

**Comment by Alexsandruss** [comment]
@SmirnovEgorRu , @ShvetsKS

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 1]
```diff
@@ -39,6 +40,59 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
...
+    IdxType * swapIdx = swapIdxArr.get();
+    DAAL_CHECK_MALLOC(swapIdx);
+
+    const size_t blockSize = 16 * BLOCK_CONST;
+    const size_t nBlocks   = n / blockSize + !!(n % blockSize);
+
+    daal::conditional_threader_for(n > THREADING_BORDER, nBlocks, [&](size_t iBlock) {
+        const size_t start = iBlock * blockSize;
+        const size_t end   = daal::services::internal::min<cpu, size_t>(start + blockSize, n);
+        daal::internal::BaseRNGs<cpu> baseRng(seed + iBlock);
```
Comment (ShvetsKS) [inline cpp/daal/src/data_management/train_test_split.cpp:61]: not sure that sequential version will reproduce the same results with such approach. (generated numbers will be  different even for number of threads changes)

please use `skipAhead` method. Details can be discussed in mail.

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 2]
```diff
@@ -39,6 +40,59 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
 const size_t BLOCK_CONST      = 2048;
 const size_t THREADING_BORDER = 8388608;
 
+template <typename IdxType, daal::CpuType cpu>
+services::Status generateShuffledIndicesImpl(const NumericTablePtr & idxTable, const unsigned int seed)
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/data_management/train_test_split.cpp:44]: Please, provide profiling for following blocks of code to understand optimizations opportunities:
1) 46-56
2) 58 - 69
3) 71 - 74
4) full function

File: cpp/daal/src/data_management/train_test_split.cpp (hunk 1 above)
Comment (SmirnovEgorRu) [inline cpp/daal/src/data_management/train_test_split.cpp:61]: Creating new RNG and Engine every time can be not cheap.

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 3]
```diff
@@ -39,6 +40,59 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
...
+    for (size_t i = 0; i < n; ++i)
+    {
+        daal::services::internal::swap<cpu, IdxType>(idx[i], idx[swapIdx[i]]);
+    }
+
+    return services::Status();
+}
+
+template <typename IdxType>
+void generateShuffledIndicesDispImpl(const NumericTablePtr & idxTable, const unsigned int seed)
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/data_management/train_test_split.cpp:80]: I suppose we can merge 2 functions:
- generateShuffledIndices
- trainTestSplit

into one function to provide more user-friendly API

File: cpp/daal/src/data_management/train_test_split.cpp (hunk 1 above)
Comment (ShvetsKS) [reply cpp/daal/src/data_management/train_test_split.cpp:61]: I made a mistake, seems as `nBlocks` doesn't depends on number of threads we will see similar sequence of generated numbers.
But we should use seed which was provided by user and do not change it.
So `skipAhead` or `leapfrog` methods are still applicable.

Comment (Alexsandruss) [reply cpp/daal/src/data_management/train_test_split.cpp]: Fixed and changed to skipAhead method.

**Comment by Alexsandruss** [comment]
@SmirnovEgorRu @ShvetsKS, pull request was updated.

File: cpp/daal/src/data_management/train_test_split.cpp (hunk 1 above)
Comment (Alexsandruss) [reply cpp/daal/src/data_management/train_test_split.cpp]: It's cheap even with RNG state setting (especially, compared to swapping).

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 4]
```diff
@@ -39,6 +41,117 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
...
+    daal::internal::RNGs<unsigned int, cpu> rng;
+
+    if (nSkip != 0) baseRng.skipAhead(nSkip);
+
+    daal::services::internal::TArray<unsigned int, cpu> tmpArr(n + 16);
+    unsigned int * tmp = tmpArr.get();
+    DAAL_CHECK_MALLOC(tmp);
+    rng.uniformBits32(n + 16, tmp, baseRng.getState());
+
+    for (size_t i = 0; i < n; ++i) randomNumbers[nSkip + i] = tmp[i];
```
Comment (ShvetsKS) [inline cpp/daal/src/data_management/train_test_split.cpp:83]: I think it would be better to generate `n` elements right to `randomNumbers` buffer without copy, and than generate 16 elements to `tmp` (tmpArr(16)). 
**And it's still not clear**: why we generates `n+16` but uses only first `n` elements?

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 5]
```diff
@@ -39,6 +41,117 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
...
+    DAAL_CHECK_MALLOC(baseRngState);
+    baseRng.saveState(baseRngState);
+    for (size_t i = 4; i < 628; ++i) baseRngState[i] = rngState[i - 4];
+    baseRngState[628] = 624;
+    baseRng.loadState(baseRngState);
+    daal::internal::RNGs<unsigned int, cpu> rng;
+
+    if (nSkip != 0) baseRng.skipAhead(nSkip);
+
+    daal::services::internal::TArray<unsigned int, cpu> tmpArr(n + 16);
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/data_management/train_test_split.cpp:78]: ```unsigned int``` can be uint16_t and uint32_t depending on 32-bit/32e-bit platforms, It can cause to wrong behavior.

File: cpp/daal/src/externals/service_rng_mkl.h [hunk 6]
```diff
@@ -33,6 +33,7 @@
 #define __DAAL_BRNG_MT19937                   VSL_BRNG_MT19937
 #define __DAAL_BRNG_MCG59                     VSL_BRNG_MCG59
 #define __DAAL_RNG_METHOD_UNIFORM_STD         VSL_RNG_METHOD_UNIFORM_STD
+#define __DAAL_RNG_METHOD_UNIFORMBITS32_STD   0
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/externals/service_rng_mkl.h:36]: Why 0?

Comment (Alexsandruss) [reply cpp/daal/src/externals/service_rng_mkl.h:36]: VSL_RNG_METHOD_UNIFORMBITS32_STD (value is 0) don't exist in mkl fpk, so, it's workaround

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 7]
```diff
@@ -39,6 +41,117 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
...
+template <daal::CpuType cpu>
+services::Status generateRandomNumbers(const int * rngState, unsigned int * randomNumbers, const size_t nSkip, const size_t n)
+{
+    daal::internal::mkl::BaseRNG<cpu> baseRng(0, VSL_BRNG_MT19937);
+    daal::services::internal::TArray<unsigned int, cpu> baseRngStateArr(baseRng.getStateSize() / 4);
+    unsigned int * baseRngState = baseRngStateArr.get();
+    DAAL_CHECK_MALLOC(baseRngState);
+    baseRng.saveState(baseRngState);
+    for (size_t i = 4; i < 628; ++i) baseRngState[i] = rngState[i - 4];
+    baseRngState[628] = 624;
```
[the remark this answers is not in the document]
Comment (Alexsandruss) [reply cpp/daal/src/data_management/train_test_split.cpp:72]: baseRngState[628] is start position of generator, need for alignment to numpy RNG

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 8]
```diff
@@ -39,6 +41,117 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
...
+    baseRng.loadState(baseRngState);
```
[the remark this answers is not in the document]
Comment (Alexsandruss) [reply cpp/daal/src/data_management/train_test_split.cpp:73]: MKL and numpy have different methods to initialize RNG from seed. So, states will be different too and manual loading is safer.

File: cpp/daal/src/data_management/train_test_split.cpp (hunk 4 above)
Comment (Alexsandruss) [reply cpp/daal/src/data_management/train_test_split.cpp:83]: It doesn't work so: if number of elements = n, then last 16 of them will be different than numpy's. Reason of that is unclear and requiring investigation with deep dive in RNG work, to save time it's better to leave code as is.

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 9]
```diff
@@ -39,6 +41,131 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
...
+
+template <daal::CpuType cpu>
+services::Status generateRandomNumbers(const int * rngState, unsigned int * randomNumbers, const size_t nSkip, const size_t n)
+{
+    // initialize baseRNG
+    daal::internal::mkl::BaseRNG<cpu> baseRng(0, VSL_BRNG_MT19937);
+    // check that it has correct size
+    DAAL_CHECK(baseRng.getStateSize() / 4 == MT19937_SIZE, daal::services::ErrorEngineNotSupported);
+    daal::services::internal::TArray<unsigned int, cpu> baseRngStateArr(baseRng.getStateSize() / 4);
+    unsigned int * baseRngState = baseRngStateArr.get();
```
Comment (ShvetsKS) [inline cpp/daal/src/data_management/train_test_split.cpp:74]: minor:
unsigned int * const baseRngState

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 10]
```diff
@@ -39,6 +41,131 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
...
+    for (size_t i = 0; i < MT19937_NUMBERS; ++i) baseRngState[i + 4] = rngState[i];
+    // change baseRNG position
+    baseRngState[MT19937_SIZE - 2] = MT19937_NUMBERS;
+    baseRng.loadState(baseRngState);
+    daal::internal::RNGs<unsigned int, cpu> rng;
+
+    if (nSkip != 0) baseRng.skipAhead(nSkip);
+
+    daal::services::internal::TArray<unsigned int, cpu> tmpArr(n + 16);
+    unsigned int * tmp = tmpArr.get();
```
Comment (ShvetsKS) [inline cpp/daal/src/data_management/train_test_split.cpp:87]: minor:
unsigned int * const

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 11]
```diff
@@ -39,6 +41,131 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
...
+
+    // check that input RNG state has needed quantity of state numbers
+    DAAL_CHECK(stateSize == MT19937_NUMBERS, daal::services::ErrorIncorrectSizeOfInputNumericTable);
+
+    daal::internal::ReadColumns<int, cpu> rngStateBlock(*rngStateTable, 0, 0, stateSize);
+    const int * rngState = rngStateBlock.get();
+    DAAL_CHECK_MALLOC(rngState);
+
+    daal::services::internal::TArray<unsigned int, cpu> randomUIntsArr(nRandomUInts);
+    unsigned int * randomUInts = randomUIntsArr.get();
```
Comment (ShvetsKS) [inline cpp/daal/src/data_management/train_test_split.cpp:120]: minor: `const`

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 12]
```diff
@@ -39,6 +41,123 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
...
+size_t genSwapIdx(size_t i, unsigned int * randomNumbers, size_t & rnIdx)
+{
+    uint32_t bitMask = i;
+    bitMask |= bitMask >> 1;
+    bitMask |= bitMask >> 2;
+    bitMask |= bitMask >> 4;
+    bitMask |= bitMask >> 8;
+    bitMask |= bitMask >> 16;
+
+    size_t j = randomNumbers[rnIdx] & bitMask;
```
Comment (PetrovKP) [inline cpp/daal/src/data_management/train_test_split.cpp:57]: minnor: Use const when you do not change data/variables

Comment (Alexsandruss) [reply cpp/daal/src/data_management/train_test_split.cpp:57]: But it actually changes

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 13]
```diff
@@ -39,6 +41,123 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
...
+
+            s |= generateRandomNumbers<cpu>(rngState, randomUInts, start, end - start);
+        });
+    }
+    else
+    {
+        s |= generateRandomNumbers<cpu>(rngState, randomUInts, 0, nRandomUInts);
+    }
+
+    for (size_t i = 0; i < n; ++i) idx[i] = i;
```
[the remark this answers is not in the document]
Comment (Alexsandruss) [reply cpp/daal/src/data_management/train_test_split.cpp:132]: Fixed

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 14]
```diff
@@ -39,6 +41,123 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
...
+    // initialize baseRNG
+    daal::internal::mkl::BaseRNG<cpu> baseRng(0, VSL_BRNG_MT19937);
+    // check that it has correct size
+    DAAL_CHECK(baseRng.getStateSize() / 4 == MT19937_SIZE, daal::services::ErrorEngineNotSupported);
+    daal::services::internal::TArray<unsigned int, cpu> baseRngStateArr(baseRng.getStateSize() / 4);
+    unsigned int * baseRngState = baseRngStateArr.get();
+    DAAL_CHECK_MALLOC(baseRngState);
+    baseRng.saveState(baseRngState);
+    // copy input state numbers to baseRNG
+    for (size_t i = 0; i < MT19937_NUMBERS; ++i) baseRngState[i + MT19937_NUMBERS_OFFSET] = rngState[i];
```
[the remark this answers is not in the document]
Comment (Alexsandruss) [reply cpp/daal/src/data_management/train_test_split.cpp:79]: Fixed

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 15]
```diff
@@ -39,6 +41,123 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
 const size_t BLOCK_CONST      = 2048;
 const size_t THREADING_BORDER = 8388608;
 
+const size_t MT19937_NUMBERS        = 624;
+const size_t MT19937_SIZE           = 631;
```
[the remark this answers is not in the document]
Comment (Alexsandruss) [reply cpp/daal/src/data_management/train_test_split.cpp:45]: MT19937_SIZE is size of MT19937 stream in 4 bytes, MT19937_NUMBERS is quantity of recurring numbers underneath generator

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 16]
```diff
@@ -39,6 +40,59 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
...
+    DAAL_CHECK_MALLOC(swapIdx);
+
+    const size_t blockSize = 16 * BLOCK_CONST;
+    const size_t nBlocks   = n / blockSize + !!(n % blockSize);
+
+    daal::conditional_threader_for(n > THREADING_BORDER, nBlocks, [&](size_t iBlock) {
+        const size_t start = iBlock * blockSize;
+        const size_t end   = daal::services::internal::min<cpu, size_t>(start + blockSize, n);
+        daal::internal::BaseRNGs<cpu> baseRng(seed + iBlock);
+        daal::internal::RNGs<IdxType, cpu> rng;
```
[the remark this answers is not in the document]
Comment (Alexsandruss) [reply cpp/daal/src/data_management/train_test_split.cpp]: In newer code MT19937 is explicitly declared, scikit-learn/numpy uses it for shuffling too

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 17]
```diff
@@ -39,6 +41,127 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
...
+    daal::internal::mkl::BaseRNG<cpu> baseRng(0, VSL_BRNG_MT19937);
+    // check that it has correct size
+    DAAL_CHECK(baseRng.getStateSize() / 4 == MT19937_SIZE, daal::services::ErrorEngineNotSupported);
+    daal::services::internal::TArray<unsigned int, cpu> baseRngStateArr(baseRng.getStateSize() / 4);
+    unsigned int * baseRngState = baseRngStateArr.get();
+    DAAL_CHECK_MALLOC(baseRngState);
+    baseRng.saveState(baseRngState);
+    // copy input state numbers to baseRNG
+    services::internal::daal_memcpy_s(baseRngState + MT19937_NUMBERS_OFFSET, MT19937_NUMBERS * sizeof(unsigned int), rngState,
+                                      MT19937_NUMBERS * sizeof(unsigned int));
```
Comment (PetrovKP) [inline cpp/daal/src/data_management/train_test_split.cpp:80]:
```suggestion
                                      MT19937_NUMBERS * sizeof(int));
```

File: cpp/daal/src/data_management/train_test_split.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (PetrovKP) [reply cpp/daal/src/data_management/train_test_split.cpp:45]: Add comment please

**Comment by PetrovKP** [comment]
http://intel-ci.intel.com/eaca1536-d628-f170-b00d-8cdcd4b723a1
